File: workspacejava/src/main/java/dom/ChemicalNumberTheory.java
package dom;



import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.math.BigInteger;
import java.util.*;

public class ChemicalNumberTheory {

    public static void main(String[] args) throws Exception {
        BufferedReader reader = new BufferedReader(
                new InputStreamReader(System.in));

        // Reading data using readLine
//        System.out.println("please input 3-10 symbols of chemical elements from professor's list separated by spaces:");
        String input = reader.readLine();
//        if (input == null || input.length() == 0 || input.trim().length() == 0) {
//            System.out.println("input is blank, exit");
//            return;
//        }
        String[] inputArray = input.split(" "); // split by space
        if (inputArray.length < 3 || inputArray.length > 10) {
            System.out.println("please input following instructions~");
            return;
        }
        // Printing the read line
        System.out.println(findHighestAffinity(inputArray));
    }

    public static Integer findHighestAffinity(String[] inputs) {
        Integer result = 0;
        Map<String, Integer> affinityMap = affinity(inputs);
        for (int i = 0; i < inputs.length; i++) {
            if (i == inputs.length - 1) {
                // skip last one
                break;
            }
            for (int j = i + 1; j < inputs.length; j++) {
                BigInteger valueI = BigInteger.valueOf(affinityMap.get(inputs[i]));
                BigInteger valueJ = BigInteger.valueOf(affinityMap.get(inputs[j]));
                result = Math.max(result, Integer.parseInt(valueI.gcd(valueJ) + ""));
//                System.out.println("value 1 " + valueI + ", value2 " + valueJ + ",gcd " + Integer.parseInt(valueI.gcd(valueJ) + ""));
            }
        }
        return result;
    }

    private static Map<String, Integer> affinity(String[] inputs) {
        Map<String, Integer> result = new HashMap<>();
        List<String> dictionary = initDictionary(); // get init dictionary
        for (int i = 0; i < inputs.length; i++) {
            String symbol = inputs[i];
            if (symbol.length() == 1 && dictionary.contains(symbol)) {
                result.put(symbol, dictionary.indexOf(symbol)); // index equal to number face value
            } else if (symbol.length() == 2) {
                String first = symbol.substring(0,1);
                String second = symbol.substring(1,2);
                if (dictionary.contains(first) && dictionary.contains(second)) {
                    Integer base = dictionary.indexOf(second) == 0 ? 2 : (dictionary.indexOf(second) + 1);
                    Integer affinity = base * dictionary.indexOf(first) + dictionary.indexOf(second);
                    result.put(symbol, affinity);
                }
            }
        }
        return result;
    }

    private static List<String> initDictionary() {
        List<String> dictionary = new ArrayList<>();
        // 0-9
        for (int i = 0; i < 10; i++) {
            dictionary.add(i + "");
        }
        // A-Z
        for (int i = 0; i < 26; i++) {
            dictionary.add(String.valueOf((char) ('a' + i)).toUpperCase());
        }
        // a-z
        for (int i = 0; i < 26; i++) {
            dictionary.add(String.valueOf((char) ('a' + i)));
        }
//        System.out.println("dictionary is: " + Arrays.toString(dictionary.toArray()));
        return dictionary;
    }


}

File: workspacejava/src/main/java/dom/ChemicalNumberTheory.py
import sys
import math

def findHighestAffinity(input) :
    inputArray = input.split()
    dict = affinity(inputArray)
    temp = 0
    for key in dict.keys():
        keys = list(dict)
        num = len(keys)
        if key == keys[num-1]:
            break
        for key_two in keys[1:num]:
            hfc = math.gcd(dict[key], dict[key_two])
            if hfc > temp:
                temp = hfc
#         if keys[0] == key:
#             temp = dict[key]
#         else:
#             temp = math.gcd(temp, dict[key])
#     for sym,aff in dict.items():
#         if dict.keys().index(sym) == 0:
#             temp = aff
#         else :
#             temp = math.gcd(temp, aff)

    return temp

def affinity (inputArray):
    dictionary = initDictionary()
    dic = {}
    for sym in inputArray:
        if len(sym) == 2:
            first = sym[0]
            second = sym[1]
            idx = dictionary.index(second)
            if idx == 0:
                base = 2
            else :
                base = idx + 1
            affinity = base * dictionary.index(first) + idx
            dic[sym] = affinity
        elif len(sym) == 1:
            dic[sym] = dictionary.index(sym)

        return dic

def initDictionary () :
    dictionary = []
    ## 0-9
    for i in (0,10):
        dictionary.append(i)
    ## A-X
        for i in range(0, 26):
           alpha = chr(ord('a') + i)
           dictionary.append(alpha.upper())
    ## a-z
    for i in range(0, 26):
       alpha = chr(ord('a') + i)
       dictionary.append(alpha)
    return dictionary

if __name__ == "__main__":
    input = sys.stdin.readline()
    print(findHighestAffinity(input))
